staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go

```diff
@@ -167,45 +170,137 @@ func (c *delegatingCRConverter) Convert(in, out, context interface{}) error {
 	if !ok {
 		// TODO: should this be a typed error?
 		return nil, fmt.Errorf("%v is unstructured and is not suitable for converting to %q", fromGVK.String(), target)
 	}
 
+	isList := false
+	var list *unstructured.UnstructuredList
+	switch t := in.(type) {
+	case *unstructured.Unstructured:
+		list = &unstructured.UnstructuredList{Items: []unstructured.Unstructured{*t}}
+	case *unstructured.UnstructuredList:
+		list = t
+		isList = true
+	default:
+		return nil, fmt.Errorf("unexpected type %T", in)
+	}
+
 	// A smoke test in API machinery calls the converter on empty objects during startup. The test is initiated here:
 	// https://github.com/kubernetes/kubernetes/blob/dbb448bbdcb9e440eee57024ffa5f1698956a054/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L201
 	if isEmptyUnstructuredObject(in) {
-		return NewNOPConverter().Convert(in, toGVK.GroupVersion())
-	}
-
-	// Special-case typed scale conversion if this custom resource supports a scale endpoint
-	if c.convertScale {
-		if _, isInScale := in.(*autoscalingv1.Scale); isInScale {
-			return typedscheme.Scheme.ConvertToVersion(in, target)
+		converted, err := NewNOPConverter().Convert(list, toGVK.GroupVersion())
+		if err != nil {
+			return nil, err
 		}
+		if !isList {
+			return &converted.Items[0], nil
+		}
+		return converted, nil
 	}
 
+	desiredAPIVersion := toGVK.GroupVersion().String()
 	if !c.validVersions[toGVK.GroupVersion()] {
-		return nil, fmt.Errorf("request to convert CR to an invalid group/version: %s", toGVK.GroupVersion().String())
+		return nil, fmt.Errorf("request to convert CR to an invalid group/version: %s", desiredAPIVersion)
 	}
 	// Note that even if the request is for a list, the GV of the request UnstructuredList is what
 	// is expected to convert to. As mentioned in the function's document, it is not expected to
 	// get a v1.List.
 	if !c.validVersions[fromGVK.GroupVersion()] {
 		return nil, fmt.Errorf("request to convert CR from an invalid group/version: %s", fromGVK.GroupVersion().String())
 	}
-	// Check list item's apiVersion
-	if list, ok := in.(*unstructured.UnstructuredList); ok {
-		for i := range list.Items {
-			expectedGV := list.Items[i].GroupVersionKind().GroupVersion()
-			if !c.validVersions[expectedGV] {
-				return nil, fmt.Errorf("request to convert CR list failed, list index %d has invalid group/version: %s", i, expectedGV.String())
-			}
+
+	var objectsToConvert []*unstructured.Unstructured
+	objectsToConvert, err := getObjectsToConvert(list, desiredAPIVersion, c.validVersions)
+	if err != nil {
+		return nil, err
+	}
+
+	objCount := len(objectsToConvert)
+	if objCount == 0 {
+		// no objects needed conversion
+		if !isList {
+			// for a single item, return as-is
+			return in, nil
+		}
+		// for a list, set the version of the top-level list object (all individual objects are already in the correct version)
+		out := list.DeepCopy()
+		out.SetAPIVersion(desiredAPIVersion)
+		return out, nil
+	}
```

Comment:
Is there an advantage to pre-checking this instead of just passing the data to the converter and letting it handle these cases?

Reply:
Code move from https://github.com/kubernetes/kubernetes/blob/4db6bde859d912e11ab081a06e8ea17b2e044f5d/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go#L251-L262

Reply:
Oops, I missed that this whole section was just a move. Please feel free to close all my comments about it.